### PR TITLE
Several changes that made TvC work more.

### PR DESCRIPTION
**Pull request by SMTDDR**

### CODE REVIEW

**Comment by SMTDDR** [comment]
Notes:
- Reduced time out - I remember it was said other games need more time, but TvC has issues as long as a "ghost" server keeps hanging around. TvC works best when servers are droppped after 1 min of silence.
- The Remove nulls commit gets undone in the "tvc leaderboard support" commit.
- For each heartbeat I put in the ingamesn from the naslogin data. That's the name of the player that shows up during battle and I need that to be able to display who is currently playing on another webpage.
- dwc_suspend = 1 was brought back because somehow an issue arose from changing it immediately based on ports not matching.

**Comment by SMTDDR** [comment]
Compressed all commits into 1.

**Comment by polaris-** [comment]
Thanks for the pull request. It'll take me a little while to go through all of this so give me some time and I'll leave some comments on it later.

File: gamespy_qr_server.py [hunk 1]
```diff
@@ -238,22 +263,26 @@ def handle_packet(self, socket, recv_data, address):
...
-                self.log(logging.DEBUG, address, "Sent client registered to %s:%s..." % (address[0], address[1]))
+                self.log(logging.DEBUG, address, session_id, "Sent client registered to %s:%s..." % (address[0], address[1]))
+
+                if self.sessions[session_id].heartbeat_data != None:
+                    self.update_server_list(session_id, self.sessions[session_id].heartbeat_data)
+
             else:
                 # Failed the challenge, request another during the next heartbeat
                 self.sessions[session_id].sent_challenge = False
+                self.server_manager.delete_server(k['gamename'] , session_id)
```
Comment (polaris-) [inline gamespy_qr_server.py:278]: See commit 888caae6102e13eb3f78d9881264c12055547fb2

File: gamespy_qr_server.py [hunk 2]
```diff
@@ -261,14 +290,23 @@ def handle_packet(self, socket, recv_data, address):
...
             for i in range(0, len(d), 2):
-                #self.log(logging.DEBUG, address, "%s = %s" % (d[i], d[i+1]))
+                #self.log(logging.DEBUG, address, session_id, "%s = %s" % (d[i], d[i+1]))
                 k[d[i]] = d[i+1]
 
+            if "gamename" in k and 'dwc_pid' in k:
+              try:
+                profile = self.db.get_profile_from_profileid(k['dwc_pid'])  
+                naslogin = self.db.get_nas_login_from_userid(profile['userid'])
+                k['ingamesn'] = str(naslogin['ingamesn'])#convert to string from unicode(which is just a base64 string anyway)
```
Comment (polaris-) [inline gamespy_qr_server.py:300]: I'm kinda curious about the performance issues involved with querying the database twice for every user every 10 seconds or so. Unless there's a need to do it for every heartbeat (maybe the user can or does frequently change their name in-game?), I think a better approach is to get the ingamesn once and cache it in the session data. I've implemented that while importing your changes.

File: gamespy_qr_server.py [hunk 3]
```diff
@@ -302,8 +340,17 @@ def handle_packet(self, socket, recv_data, address):
...
                             time.sleep(0.5)
 
+            if 'publicip' in k and k['publicip'] == "0": #and k['dwc_hoststate'] == "2": # When dwc_hoststate == 2 then it doesn't send an IP, so calculate it ourselves
+                be = self.sessions[session_id].console != 0
+                k['publicip'] = str(utils.get_ip(bytearray([int(x) for x in address[0].split('.')]), 0, be))
 
-            if self.sessions[session_id].sent_challenge == False:
+            if 'dwc_suspend' in k and k['dwc_suspend'] == "1":
+                self.log(logging.DEBUG, address, session_id, "dwc_suspend=1, settiing publicport to %s..." % (str(address[1])))
+                k['publicport'] = str(address[1])
```
Comment (polaris-) [inline gamespy_qr_server.py:349]: See commit 8ddc427d69d384477e85d2510d201655c3d0d0c4 (Lines 338-341)

**Comment by polaris-** [comment]
I'm not going to accept the changes to nas_server and I'll leave tatvscapwii_gamestats2 in your branch for now. I don't know how we'll handle game-specific gamestats servers in the server code yet so I'll leave that for the future.

**Pull request closed without merging**